### DISCUSSION TITLE
MetaMask as an Identity provider for keycloak

### DISCUSSION THREAD

**Discussion by gkozyryatskyy** [keycloak/keycloak#16799] [Q&A] [answered]
Hi guys!
I wonder, is there a possibility to use MetaMask based authentication in keycloak?
Here is a basic idea of this kind of authentication https://www.toptal.com/ethereum/one-click-login-flows-a-metamask-tutorial

If there is no such thing implemented, any ideas how i can implement it, for example, as new Identity provider?
From code side this is easy, we basically just need to validate message signed by MetaMask plugin. But how to add it to keycloak?

**Comment by gkozyryatskyy** [comment, marked as answer]
If someone will need this, I just end up using hosted OIDC with eth wallets support https://docs.login.xyz/servers/oidc-provider/hosted-oidc-provider. This OIDC I added as an Identity providers OIDC integration on keycloak.

**Comment by aliIsazadeh** [comment]
For anyone still landing here from search: I ended up building this as a proper Authenticator SPI plugin — Sign-In with Ethereum (SIWE / EIP-4361) and Solana (SIWS), for both EOA and smart-contract wallets (EIP-1271, plus EIP-6492 counterfactual signatures), running as one authenticator inside your existing browser flow. No second OIDC server to federate, no wallet SDK on the backend, and no new infrastructure — the single-use nonce lives in Keycloak's own authentication-session notes.

- Repo (Apache-2.0): https://github.com/aliIsazadeh/keycloak-wallet-authenticator
- Latest release (CI-built JAR + SHA-256 checksum): https://github.com/aliIsazadeh/keycloak-wallet-authenticator/releases/latest
- Design write-up (domain binding, nonce handling, EIP-1271/6492 without web3j): https://dev.to/aliisazadeh/building-a-production-grade-wallet-login-authenticator-for-keycloak-ai

It's tested end-to-end against a real Keycloak 25 in Testcontainers and verified with real MetaMask and Phantom. It's early — if you try it and something breaks, an issue on the repo would be very welcome.
